components/brave_wayback_machine/brave_wayback_machine_tab_helper.cc

```diff
@@ -5,44 +5,93 @@
 #include "content/public/browser/browser_context.h"
 #include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/storage_partition.h"
 #include "content/public/browser/web_contents.h"
 #include "net/http/http_response_headers.h"
 #include "net/http/http_status_code.h"
+#include "ui/views/widget/widget.h"
+
+// static
+void BraveWaybackMachineTabHelper::CreateIfNeeded(
+    content::WebContents* web_contents) {
+  if (base::CommandLine::ForCurrentProcess()->HasSwitch(
+          switches::kDisableBraveWaybackMachineExtension)) {
+    return;
+  }
+
+  BraveWaybackMachineTabHelper::CreateForWebContents(web_contents);
+}
 
 BraveWaybackMachineTabHelper::BraveWaybackMachineTabHelper(
     content::WebContents* contents)
     : WebContentsObserver(contents),
       content::WebContentsUserData<BraveWaybackMachineTabHelper>(*contents),
       pref_service_(*user_prefs::UserPrefs::Get(contents->GetBrowserContext())),
-      weak_factory_(this) {}
+      wayback_machine_url_fetcher_(
+          this,
+          contents->GetBrowserContext()
+              ->GetDefaultStoragePartition()
+              ->GetURLLoaderFactoryForBrowserProcess()) {}
+
+BraveWaybackMachineTabHelper::~BraveWaybackMachineTabHelper() {
+  CHECK(!wayback_state_changed_callback_);
+  CloseWindowIfPossible();
+}
+
+void BraveWaybackMachineTabHelper::FetchWaybackURL() {
+  SetWaybackState(WaybackState::kFetching);
+  wayback_machine_url_fetcher_.Fetch(web_contents()->GetVisibleURL());
+}
 
-BraveWaybackMachineTabHelper::~BraveWaybackMachineTabHelper() = default;
+void BraveWaybackMachineTabHelper::SetWaybackStateChangedCallback(
+    WaybackStateChangedCallback callback) {
+  // callback should be set only once.
+  // And it should be cleared only when there is existing one.
+  if (callback) {
+    CHECK(!wayback_state_changed_callback_);
+  } else {
+    CHECK(wayback_state_changed_callback_);
+  }
 
-void BraveWaybackMachineTabHelper::set_delegate(
-    std::unique_ptr<BraveWaybackMachineDelegate> delegate) {
-  DCHECK(delegate);
-  delegate_ = std::move(delegate);
+  wayback_state_changed_callback_ = std::move(callback);
 }
 
-void BraveWaybackMachineTabHelper::DidFinishNavigation(
+// When wayback url is loaded via user action,
+// we don't want to clear wayback state.
+void BraveWaybackMachineTabHelper::DidStartNavigation(
     content::NavigationHandle* navigation_handle) {
-  DCHECK(delegate_);
-  if (!IsWaybackMachineEnabled())
+  if (!navigation_handle->IsInMainFrame() ||
+      navigation_handle->IsSameDocument()) {
+    return;
+  }
+
+  // Don't reset current state if it's wayback url navigation.
+  // Otherwise, we lost kLoaded state after loading it.
+  if (wayback_url_navigation_ &&
+      wayback_url_navigation_.get() == navigation_handle) {
     return;
+  }
+
+  // Reset to initial state whenever new navigation starts.
+  SetWaybackState(WaybackState::kInitial);
+}
 
+void BraveWaybackMachineTabHelper::DidFinishNavigation(
```

Comment:
If you feel like this should have more checks like `NavigationHandle::IsInPrimaryMainFrame && !NavigationHandle::IsSameDocument && NavigationHandle::HasCommitted` here, PrimaryFrameChanged() might be a better choice.

But I think we should check "IsInPrimaryMainFrame()` at least.

Reply:
Updated to use `IsInPrimaryMainFrame()`.